src/librustc_typeck/check/pat.rs

```diff
@@ -380,9 +380,38 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 E0029,
```

Comment:
At this point the block of the `if !lhs_compat || !rhs_compat ` should be moved to a separate function to avoid cluttering the happy path.